### PR TITLE
auto-epp: init at 1.2.1

### PR DESCRIPTION
**Pull request by lamarios**
nixos/services/hardware/auto-epp-go: init

module to set up auto-epp-go with its kernel parameter, systemd service and config file

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:7]: Consider making this an [`mkPackageOption`](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.options.mkPackageOption)

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:24]: Although there are just 2 settings at the moment, I'd still recommend following RFC 42 with `format = pkgs.formats.ini { };`

```suggestion
      settings = types.submodule {
        freeformType = format.type;
        options = {
          settings = {
            epp_state_for_AC = mkOption {
              type = types.str;
              default = "power";
              description = lib.mdDoc ''
                energy_performance_preference when on battery

								::: {.note}
								See available epp states by running:
								{command}`cat /sys/devices/system/cpu/cpu0/cpufreq/energy_performance_available_preferences`
								:::
              '';
            };

            epp_state_for_BAT = mkOption {
              type = types.str;
              default = "balance_performance";
              description = lib.mdDoc ''
                `energy_performance_preference` when plugged in

								::: {.note}
								See available epp states by running:
								{command}`cat /sys/devices/system/cpu/cpu0/cpufreq/energy_performance_available_preferences`
								:::
              '';
            };
          };
        };
      };
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:44]:
```suggestion
    systemd.services.auto-epp = mkIf cfg.enable {
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:46]: Does this service really need networking in order to work?

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:50]: Does this service need to run as root? Maybe we could make it `DynamicUser` and give it the permissions it needs through `CapabilityBoundingSet`?

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:22]:
```suggestion
    mainProgram = pname;
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:40]: RFC 42 compliance part 2: this can now become

```suggestion
    environment.etc."auto-epp.conf".source = format.generate "auto-epp.conf" cfg.settings;
```

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:18]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    install -Dm555 auto-epp $out/bin/auto-epp
    
    runHook postInstall
  '';
```

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:1]: Arguments have a recommended order: https://nixos.org/manual/nixpkgs/stable/#sec-syntax (last point)

```suggestion
{ lib, stdenv, fetchFromGitHub, python3 }:
```

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:14]: I'm not entirely convinced this is supposed to be a normal `buildInputs`. I think maybe it should be wrapped somehow?

```suggestion
  buildInputs = [ python3 ];
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:58]:
```suggestion
  };
  
  meta.maintainers = with lib.maintainers; [ lamarios ];
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:2]: There are ongoing efforts to avoid `with lib;` statements. I won't block the PR on this, but would you mind changing it out for an `inherit (lib) ...;` for the most commonly used items, and just prepending `lib.` to everything that's only used once or twice?

**Review by h7x4 — CHANGES_REQUESTED** [review]
You'll have to split up your changes into three commits:

```
maintainers: add lamarios
auto-epp: init at 1.2.1
nixos/auto-epp: init
```

I'm not entirely convinced that the packaging is correct, but I'll leave that to other reviewers who might know better than me.

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [reply nixos/modules/services/hardware/auto-epp.nix]: I tried this on my machines and it is not working, i tried the following:
```nix
services.auto-epp = {
     enable = true;
     settings.epp_state_for_AC = "performance";
  };
# or
services.auto-epp = {
     enable = true;
     settings.settings.epp_state_for_AC = "performance";
  };
# or
services.auto-epp = {
     enable = true;
     settings.epp_state_for_AC = "performance";
  };
```
I always get that the option does not exist. 

I managed to make it work with:
```nix
services.auto-epp = {
      enable = mkEnableOption (lib.mdDoc "auto-epp for amd active pstate");
     # settings = lib.mkOption {
        type = lib.types.submodule {
          freeformType = format.type;
          options = {
            settings.epp_state_for_AC = mkOption {
        ....
```

but then I need  to set i t up this way:
```nix
services.auto-epp = {
     enable = true;
     settings.settings.epp_state_for_AC = "performance";
  };
```

which looks silly

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [reply nixos/modules/services/hardware/auto-epp.nix:67]: I'll need to research on this as i have no idea on how those work.

Comment (lamarios) [reply nixos/modules/services/hardware/auto-epp.nix:72]: After some check, it requires to run as root, logs from journalctl:
```
Oct 17 12:02:01 yoga7 auto-epp[53340]: auto-epp must be run with root privileges.
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [reply nixos/modules/services/hardware/auto-epp.nix]: ok, all addressed except commit splitting, let's see if it's good to go first then i'll split

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [reply nixos/modules/services/hardware/auto-epp.nix]: Cool. Maybe we should notify upstream

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:75]:
```suggestion
        ExecStart = lib.getExe cfg.package;
```

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:23]: nit: while i understand the intent here, it's semantically wrong. The executable doesn't necessarily have the same name as the package - it just happens to be the case here (and for many other packages as well).

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [inline pkgs/tools/system/auto-epp/default.nix:1]: Would you be willing to move your package to the new `by-name` set by the way? AFAIK it's not obligatory just yet, but many new packages are shifting towards using this.

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:65]: nit:

```suggestion

```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:55]: nit: fixed some odd-even indentation issues, added a missing `mdDoc` and removed some newlines
```suggestion
  options = {
    services.auto-epp = {
      enable = lib.mkEnableOption (lib.mdDoc "auto-epp for amd active pstate");

      package = lib.mkPackageOptionMD pkgs "auto-epp" {};

      settings = mkOption {
        type = types.submodule {
          freeformType = format.type;
          options = {
            Settings = {
              epp_state_for_AC = mkOption {
                type = types.str;
                default = "power";
                description = lib.mdDoc ''
                  energy_performance_preference when on battery

                  ::: {.note}
                  See available epp states by running:
                  {command}`cat /sys/devices/system/cpu/cpu0/cpufreq/energy_performance_available_preferences`
                  :::
                '';
              };

              epp_state_for_BAT = mkOption {
                type = types.str;
                default = "balance_performance";
                description = lib.mdDoc ''
                  `energy_performance_preference` when plugged in

                  ::: {.note}
                  See available epp states by running:
                  {command}`cat /sys/devices/system/cpu/cpu0/cpufreq/energy_performance_available_preferences`
                  :::
                '';
              };
            };
          };
        };
        default = {};
        description = lib.mdDoc "Settings for the auto-epp application";
      };
    };
  };
```

**Review by h7x4 — COMMENTED** [review]
Okay, just a few more comments

File: pkgs/tools/system/auto-epp/default.nix
Comment (lamarios) [reply pkgs/tools/system/auto-epp/default.nix:1]: sure, any docs or example i can read ?

File: pkgs/tools/system/auto-epp/default.nix
Comment (lamarios) [reply pkgs/tools/system/auto-epp/default.nix:23]: sure, i'll hardcode it

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [reply nixos/modules/services/hardware/auto-epp.nix]: any linter i can use that follows nixpkgs'  requirements ?

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [reply pkgs/tools/system/auto-epp/default.nix:1]: Sure! There's https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [reply nixos/modules/services/hardware/auto-epp.nix]: While there are linters, I don't think this comment is a requirement for the most part. I know some people use [`nixpkgs-fmt`](https://nix-community.github.io/nixpkgs-fmt/)

File: pkgs/tools/system/auto-epp/default.nix
Comment (h7x4) [reply pkgs/tools/system/auto-epp/default.nix]: While you already changed this, I thought I'd just clarify that I learnt the other day that this is no longer the case. I was looking at the stable manual, but the unstable manual (and now part of the CONTRIBUTING.md docs), this recommendation has been explicitly removed

File: pkgs/tools/system/auto-epp/default.nix
Comment (lamarios) [reply pkgs/tools/system/auto-epp/default.nix:1]: if i understand properly, i just need to move my default.nix file  to  pkgs/by-name/au/auto-epp/package.nix  and remove it from the top-level/all-packages.nix ?

Comment (h7x4) [reply pkgs/tools/system/auto-epp/default.nix:1]: Yes, I think so

Comment (lamarios) [reply pkgs/tools/system/auto-epp/default.nix]: ok, done

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:64]: This is redundant, I think: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/systemd-unit-options.nix#L37-L38

**Comment by h7x4** [comment]
This is looking pretty good. You'll have to rebase your commits into

```
maintainers: add lamarios
auto-epp: init at 1.2.1
nixos/auto-epp: init
```

But apart from that, this is ready to merge as far as I'm concerned

**Comment by lamarios** [comment]
@h7x4  all done ! Thanks for the review

**Review by h7x4 — APPROVED** [review]

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [inline nixos/modules/services/hardware/auto-epp.nix:22]: @h7x4  had to make changes, i realized i inverted the defaults for battery and ac modes...

File: nixos/modules/services/hardware/auto-epp.nix
Comment (lamarios) [inline nixos/modules/services/hardware/auto-epp.nix:35]: @h7x4  had to make changes, i realized i inverted the defaults for battery and ac modes...

File: pkgs/by-name/au/auto-epp/package.nix
Comment (h7x4) [inline pkgs/by-name/au/auto-epp/package.nix:25]: I think extracting the name of the executable into it's own assignment is a bit unnecessary. It's not really something that should varies a lot, nor is it set conditionally.

```suggestion
stdenv.mkDerivation rec {
  pname = "auto-epp";
  version = "1.2.1";

  src = fetchFromGitHub {
    owner = "jothi-prasath";
    repo = "auto-epp";
    rev = "refs/tags/v${version}";
    hash = "sha256-7sI8K+7ZAdzBN/XOzYQQZ1f9t+fFo6fcXYzX6abNyQ8=";
  };

  buildInputs = [ python3 ];

  installPhase = ''
    runHook preInstall
    install -Dm555 auto-epp $out/bin/auto-epp
    runHook postInstall
  '';

  meta = with lib; {
    mainProgram = "auto-epp";
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (h7x4) [inline nixos/modules/services/hardware/auto-epp.nix:63]: This is already checked at `config = lib.mkIf cfg.enable`, no?

```suggestion
    systemd.services.auto-epp = {
```

**Review by h7x4 — COMMENTED** [review]
Some final comments

**Comment by lamarios** [comment]
> Some final comments

all done, thanks

File: nixos/modules/services/hardware/auto-epp.nix
Comment (pbsds) [inline nixos/modules/services/hardware/auto-epp.nix:64]:
```suggestion
      after = [ "multi-user.target" ];
      wantedBy  = [ "multi-user.target" ];
```

File: nixos/modules/services/hardware/auto-epp.nix
Comment (pbsds) [inline nixos/modules/services/hardware/auto-epp.nix:73]:
```suggestion
```

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 261378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>auto-epp</li>
  </ul>
</details>


I also built the manual with `nix build github:nixos/nixpkgs/refs/pull/261378/head#htmlDocs.nixosManual`.
Couple of nits, yet to test building the module.

**Comment by lamarios** [comment]
> Result of `nixpkgs-review pr 261378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 2 packages blacklisted:
> 
>     * nixos-install-tools
> 
>     * tests.nixos-functions.nixos-test
> 
> 
> 1 package built:
> 
>     * auto-epp
> 
> 
> I also built the manual with `nix build github:nixos/nixpkgs/refs/pull/261378/head#htmlDocs.nixosManual`. Couple of nits, yet to test building the module.

After committing the suggestions, there's an issue with build manual:
```
    3| let
       >             4|   cfg = config.services.auto-epp;
       >              |         ^
       >             5|   format = pkgs.formats.ini {};
       > Cacheable portion of option doc build failed.
       > Usually this means that an option attribute that ends up in documentation (eg `default` or `description`) depends on the restricted module arguments `config` or `pkgs`.

```

Looks like we can't use config.xxx in the docs. Should we hardcode the url to point to the github repo instead ?

**Comment by lamarios** [comment]
> > Should we hardcode the url to point to the github repo instead ?
> 
> Yes, I think that would be good. Most module docs point to the projects main branch, so maybe `https://github.com/jothi-prasath/auto-epp/blob/master/sample-auto-epp.conf`?

done

**Pull request merged**
